### PR TITLE
VIP only with no backend service 

### PR DESCRIPTION
**Pull request by nafets227**
This Pull Requests enables to let run Keepalived with only attaching the VIP to the node. Then e.g. HAProxy can be run via Daemonset to make use of it.

I use it in order to Forward SMTP via Proxy protocol to postfix. With normal forwarding postfix will njever get the real IP address of the caller.

### CODE REVIEW

File: keepalived-vip/controller.go [hunk 1]
```diff
@@ -171,6 +172,23 @@ func (ipvsc *ipvsControllerController) getServices(cfgMap *api.ConfigMap) []vip
 	// k -> IP to use
 	// v -> <namespace>/<service name>:<lvs method>
 	for externalIP, nsSvcLvs := range cfgMap.Data {
+		if nsSvcLvs == "" {
+			// if targtet is empty string we will not forward to any service but
+			// instead just configure the IP on the machine and let it up to 
+			// other Pods (to be started separately, e.g. by DaemonSet) like HAProxy
+			// to listen on that IP adress.
```
Comment (bowei) [inline keepalived-vip/controller.go]: targtet -> target
adress -> address

File: keepalived-vip/controller.go [hunk 2]
```diff
@@ -171,6 +172,23 @@ func (ipvsc *ipvsControllerController) getServices(cfgMap *api.ConfigMap) []vip
 	// k -> IP to use
 	// v -> <namespace>/<service name>:<lvs method>
 	for externalIP, nsSvcLvs := range cfgMap.Data {
+		if nsSvcLvs == "" {
+			// if targtet is empty string we will not forward to any service but
+			// instead just configure the IP on the machine and let it up to
```
Comment (bowei) [inline keepalived-vip/controller.go]: ... let it be up to another Pod or daemon bind to the IP address

File: keepalived-vip/controller.go [hunk 3]
```diff
@@ -171,6 +172,23 @@ func (ipvsc *ipvsControllerController) getServices(cfgMap *api.ConfigMap) []vip
...
+			// to listen on that IP adress.
+			svcs = append(svcs, vip{
+				Name:      "",
+				IP:        externalIP,
+				Port:      0,
+				LVSMethod: "VIP",
+				Backends:  nil,
+				Protocol:  "TCP",
+			})
+			glog.V(2).Infof("found service: VIP %v", externalIP)
```
Comment (bowei) [inline keepalived-vip/controller.go]: Infof("Adding VIP only service: %v", externalIP)

File: keepalived-vip/controller.go [hunk 4]
```diff
@@ -284,12 +302,13 @@ func (ipvsc *ipvsControllerController) Stop() error {
 }
 
 // newIPVSController creates a new controller from the given config.
-func newIPVSController(kubeClient *unversioned.Client, namespace string, useUnicast bool, configMapName string, vrid int) *ipvsControllerController {
+func newIPVSController(kubeClient *unversioned.Client, namespace string, useUnicast bool, configMapName string, vrid int, dryrun bool) *ipvsControllerController {
```
Comment (bowei) [inline keepalived-vip/controller.go]: Please find a different way to mock the IPVS controller, rather than interleaving the logic as a "dryrun" flag

Comment (nafets227) [inline keepalived-vip/controller.go]: I was thinking on that. I thought better to have one flag than to have a common global shared object or flag for all other objects.
Only Idea I had was another method setDryRun(), but this Comes to late.

Do you have a better idea?

**Comment by nafets227** [comment]
Updated Source Code, rewrote the Log and exluded the dry-run, since it seems not mature enough yet.

**Review by bowei — APPROVED** [review]

**Comment by nafets227** [comment]
#2630

**Comment by nafets227** [comment]
#2630

**Comment by nafets227** [comment]
@bowei what´s missing to merge this?

**Review by bowei — APPROVED** [review]

**Comment by bowei** [comment]
submit robot seems to be confused

**Comment by bowei** [comment]
@ixdy do you how to unwedge this PR?

**Comment by bowei** [comment]
@ixdy seems to still be stuck?

**Comment by spiffxp** [comment]
@nafets227 try editing the PR description to include a link to the issue you're fixing, I've had the bot react to that in the past

**Comment by cjwagner** [comment]
Someone from the OWNERS file has to comment with "/approve no-issue" @bowei 's comment didn't work because it has a leading space (" /approve no-issue").

**Comment by bowei** [comment]
bot needs to trim whitespace...

**Pull request merged**